apps/bluetooth/js/transfer.js

```diff
@@ -195,36 +196,89 @@ navigator.mozL10n.once(function showPanel() {
+              transferred();
+            }
+          };
 
-        getRequest.onsuccess = function() {
-          defaultAdapter.sendFile(targetDevice.address, getRequest.result);
-          var msg = 'getFile succeed & file is sending...';
+          getRequest.onerror = function() {
+            defaultAdapter.sendFile(targetDevice.address, blob);
+            var msg = 'getFile failed so that blob is sending without ' +
+                      'filename ' + getRequest.error && getRequest.error.name;
+            debug(msg);
+            if (--numberOfTasks === 0) {
+              transferred();
+            }
+          };
+        } else {
+          // There is no filepath, apply approach 2.
+          // Wrap the file with filename for sending.
+          // And the filename is better coming from the requsting app.
+          var TMPDIR = '.bluetooth/tmp';
+          var filename = TMPDIR + '/' + activity.source.data.filenames[index];
+          msg = 'add filename for blob, name = ' + filename;
           debug(msg);
-          if (--numberOfTasks === 0) {
-            transferred();
-          }
-        };
 
-        getRequest.onerror = function() {
-          defaultAdapter.sendFile(targetDevice.address, blob);
-          var msg = 'getFile failed so that blob is sending without filename ' +
-                    getRequest.error && getRequest.error.name;
-          debug(msg);
-          if (--numberOfTasks === 0) {
-            transferred();
-          }
-        };
+          var addNamedRequest = storage.addNamed(blob, filename);
+          addNamedRequest.onsuccess = function() {
+            var getRequest = storage.get(addNamedRequest.result);
+            getRequest.onsuccess = function() {
+              defaultAdapter.sendFile(targetDevice.address, getRequest.result);
+              var msg = 'add name in file succeed & file is sending...';
+              debug(msg);
+              if (--numberOfTasks === 0) {
+                transferred();
+              }
+            };
+
+            getRequest.onerror = function() {
+              defaultAdapter.sendFile(targetDevice.address, blob);
+              var msg = 'getFile failed after added name succeed so that ' +
+                        'blob is sending without filename ' +
+                        getRequest.error && getRequest.error.name;
+              debug(msg);
+              if (--numberOfTasks === 0) {
+                transferred();
+              }
+            };
+          };
+
+          addNamedRequest.onerror = function() {
+            console.warn('Unable to write the file: ' + this.error);
+            defaultAdapter.sendFile(targetDevice.address, blob);
+            var msg = 'add name in file failed so that blob is sending ' +
+                      'without filename ' + this.error && this.error.name;
+            debug(msg);
+            if (--numberOfTasks === 0) {
+              transferred();
+            }
+          };
+        }
```

Comment:
There is lots of repeated code that could be factored out into a utility function to call sendFile(), print the debug message, decrement numberOfTasks, etc.